sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala

```diff
@@ -106,13 +110,36 @@ abstract class PartitioningAwareFileIndex(
       PartitionDirectory(InternalRow.empty, allFiles()
         .filter(f => isNonEmptyFile(f) && matchFileMetadataPredicate(f))) :: Nil
     } else {
-      if (recursiveFileLookup) {
-        throw new IllegalArgumentException(
-          "Datasource with partition do not allow recursive file loading.")
```

Comment:
This was added in https://github.com/apache/spark/pull/24830. cc @WeichenXu123 @cloud-fan @gengliangwang FYI